CHANGELOG.md

```diff
@@ -10,6 +8,11 @@ librdkafka v1.7.1 is a maintenance release.
    connection goes down, only when all standard named brokers have been tried.
    This fixes the issue with `ERR__ALL_BROKERS_DOWN` being triggered on
    `consumer_close()`.
+ * `rd_kafka_query_watermark_offsets()`, `rd_kafka_offsets_for_times()`,
+   `consumer_lag` metric, and `auto.offset.reset` now honour
+   `isolation.level` and will return the Last Stable Offset (LSO)
```

Comment:
The problem with doing this for `rd_kafka_query_watermark_offsets` is the word `watermark` in the name. have you considered adding a new method instead (and keeping watermark offsets returning watermark offsets, but marking it as depreciated maybe?). i know the difference is probably harmless in any case here that you (or i) can think of, but preciseness is important. if someone reads the docs and see that the naming is deceptive, it will reflect poorly on the API - they will wonder what else is deceptive etc. that alone is enough reason to be precise IMO.

Reply:
I wonder if it would be beneficial to follow the Java API terminology here. They differentiate between LSO, high watermark (aka last replicated offset) and endOffset (the logical end offset whose value may be LSO or high watermark depending on isolation level).

https://kafka.apache.org/22/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#endOffsets-java.util.Collection-java.time.Duration-

> Get the end offsets for the given partitions. In the default read_uncommitted isolation level, the end offset is the high watermark (that is, the offset of the last successfully replicated message plus one). For read_committed consumers, the end offset is the last stable offset (LSO), which is the minimum of the high watermark and the smallest offset of any open transaction. Finally, if the partition has never been written to, the end offset is 0.

Reply:
query_watermark_offsets() predates the Java endOffsets() and beginningOffsets() APIs, and covers the functionality of both.

We could add endOffsets() and beginningOffsets() to match the Java client and leave query_watermark_offsets() as-is (non-isolation.level-caring), which would be straight forward but require work in all the daughter clients.
Or we could simply document the new behaviour, which would make query_watermark_offsets() do the right thing for existing applications that just care about committed messages.
I'm reasoning that if you configure isolation.level you want that worldview in all APIs. I.e., consumer lag really shouldn't include non-(txn)committed messages.

I'm not sure which is best.